File: Binary Search/binary_search.js
function binarySearch(array, key) {
    let left = 0;
    let right = array.length - 1;
    while (left <= right) {
        const mid = left + Math.floor((right - left) / 2);
        if (array[mid] === key) {
            return mid;
        }
        if (array[mid] < key) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

function recursiveSearch(array,target,start=0, end=null){
    if(end==null){
        end = array.length;
    }

    if(start > end){
        return -1;
    }
    const midpoint = Math.floor((start + end)/2);

    if(array[midpoint] == target){
        return midpoint;
    }
    else if(array[midpoint] < target){
       recursiveSearch(array,target,midpoint+1, end) 
    }
    else{
        recursiveSearch(array,target, start, midpoint-1)
    }
}

File: Binary Search/binary_search.java
public class IterativeBinarySearch {
    public static Integer binarySearch(int[] input, int target) {
        int first = 0;
        int last = input.length - 1;

        while (first <= last) {
            int mid = (first + last) / 2;

            if (input[mid] == target) {
                return mid;
            } else if (input[mid] < target) {
                first = mid + 1;
            } else {
                last = mid - 1;
            }
        }

        return null;
    }
}

public class RecursiveBinarySearch {
    public static int binarySearch(int[] input, int target, int start, int end) {
        if (start >= end) {
            return -1;
        } else {
            int mid = start + (end - start) / 2;

            if (target < input[mid]) {
                return binarySearch(input, target, start, mid-1);
            } else if (target > input[mid]) {
                return binarySearch(input, target, mid+1, end);
            } else {
                return mid;
            }
        }
    }
}